specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,246 @@
+ETH 2.0 Networking Spec - RPC Interface
+===
+
+# Abstract
+
+The Ethereum 2.0 networking stack uses two modes of communication: a broadcast protocol that gossips information to interested parties via GossipSub, and an RPC protocol that retrieves information from specific clients. This specification defines the RPC protocol.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL", NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+# Dependencies
+
+This specification assumes familiarity with the [Messaging](./messaging.md), [Node Identification](./node-identification), and [Beacon Chain](../core/0_beacon-chain.md) specifications.
+
+# Specification
+
+## Message Schemas
+
+Message body schemas are notated like this:
+
+```
+(
+    field_name_1: type
+    field_name_2: type
+)
+```
+
+SSZ serialization is field-order dependent. Therefore, fields MUST be  encoded and decoded according to the order described in this document. The encoded values of each field are concatenated to form the final encoded message body. Embedded structs are serialized as Containers unless otherwise noted.
+
+All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.
+
+## `libp2p` Protocol Names
+
+A "Protocol Name" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. A client's supported protocol paths are negotiated by the `libp2p` stack at connection time; as such they are not part of individual message bodies.
+
+## RPC-Over-`libp2p`
+
+To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:
+
+```
+(
+    id: uint64
+    method_id: uint16
+    body: Request
+)
+```
+
+and their corresponding responses are wrapped in a "response" structure:
+
+```
+(
+    id: uint64
+    result: Response
+)
+```
+
+If an error occurs, a variant of the response structure is returned:
+
+```
+(
+    id: uint64
+    error: (
+        code: uint16
+        data: bytes
+    )
+)
+```
+
+The details of the RPC-Over-`libp2p` protocol are similar to [JSON-RPC 2.0](https://www.jsonrpc.org/specification). Specifically:
+
+1. The `id` member is REQUIRED.
+2. The `id` member in the response MUST be the same as the value of the `id` in the request.
```

Comment:
Ideally we'd define the properties of this `id`. Is it monotonically increasing? Unique within a session?

Reply:
According to the JSON-RPC spec it need only be unique within a session. I'll update the doc to reflect this requirement and provide some `RECOMMENDED` options.

Having an `id` parameter like this is important since it allows RPC messages to be pipelined.

Reply:
I've updated this. See latest changes.